compiler/AST/AstToText.cpp

```diff
@@ -0,0 +1,1318 @@
+
+    if (argSym->typeExpr)
+    {
+      BlockStmt* bs = argSym->typeExpr;
+
+      if (bs->body.length == 1)
+      {
+        Expr* expr = bs->body.only();
+
+        if (UnresolvedSymExpr* sel = toUnresolvedSymExpr(expr))
+        {
+          mText += sel->unresolved;
+        }
+
+        else if (SymExpr* sel = toSymExpr(expr))
+        {
+          if (TypeSymbol* typeSym = toTypeSymbol(sel->var))
+          {
+            appendExpr(typeSym->name);
+          }
+          else
+          {
+            INT_ASSERT(false);
+          }
+        }
+
+        else
+        {
+          INT_ASSERT(false);
+        }
+      }
+
+      else
+      {
+        INT_ASSERT(false);
+      }
+    }
+
+    else if (PrimitiveType* type = toPrimitiveType(argSym->type))
+      appendExpr(type->symbol->name);
+
+    else if (AggregateType* type = toAggregateType(argSym->type))
+      appendExpr(type->symbol->name);
+
+    else
+      INT_ASSERT(false);
+  }
+  else
+  {
+    INT_ASSERT(false);
+  }
+}
+
+/************************************ | *************************************
+*                                                                           *
+* Append a normalized version of the formals list to the buffer.            *
+*                                                                           *
+************************************* | ************************************/
+
+void AstToText::appendFormals(FnSymbol* fn)
+{
+  int  count = numFormals(fn);
+  bool skip  = skipParens(fn);
+  bool first = true;
+
+  if (skip == false)
+    mText += '(';
+
+  for (int index = 1; index <= count; index++)
+  {
+    ArgSymbol* arg = formalGet(fn, index);
+
+    if (arg->hasFlag(FLAG_IS_MEME) == false)
+    {
+      if (first == true)
+      {
+        if (skip == true)
+          mText += " ";
+
+        first = false;
+      }
+      else
+        mText += ", ";
+
+      appendFormal(fn, index);
+    }
+  }
+
+  if (skip == false)
+    mText += ')';
+}
+
+bool AstToText::skipParens(FnSymbol* fn) const
+{
+  bool retval = false;
+
+  if (fn->hasFlag(FLAG_NO_PARENS))
+    retval = true;
+
+  else if (fn->hasFlag(FLAG_TYPE_CONSTRUCTOR) && fn->numFormals() == 0)
+    retval = true;
+
+  else if (fn->hasFlag(FLAG_MODULE_INIT)      && developer        == false)
+    retval = true;
+
+  else
+    retval = false;
+
+  return retval;
+}
+
+/************************************ | *************************************
+*                                                                           *
+* Append a normalized version of a formal to the buffer.                    *
+*                                                                           *
+************************************* | ************************************/
+
+// Excludes "_mt" (method token) and "this" when they are present
+int AstToText::numFormals(FnSymbol* fn) const
+{
+  int retval = fn->formals.length;
+
+  if (fn->isMethod() == true)
+    retval = retval - indexForThis(fn);
```

Comment:
This assumes that _mt and this will be both present if one is, grouped together and the first two arguments as inserted.  I hope this is true, but I'm not sure a method to get the number of formals should rely on these details.